File: Repairs/app/src/main/java/com/rooio/repairs/Registration.java
package com.rooio.repairs;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.appcompat.app.ActionBar;
import androidx.arch.core.util.Function;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.regex.Pattern;

public class Registration extends RestApi {

    private EditText firstname;
    private EditText lastname;
    private EditText email;
    private EditText password;
    private EditText restaurantname;
    private Button registerButton;
    private Button cancelButton;
    private TextView error;
    private Integer industry_int;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_registration);
        getSupportActionBar().setDisplayOptions(ActionBar.DISPLAY_SHOW_CUSTOM);
        getSupportActionBar().setCustomView(R.layout.action_bar);
        getSupportActionBar().setElevation(0);

        registerButton = findViewById(R.id.register);
        cancelButton = findViewById(R.id.cancelRegistration);
        firstname = findViewById(R.id.firstname);
        lastname = findViewById(R.id.lastname);
        email = findViewById(R.id.email);
        password = findViewById(R.id.password);
        restaurantname = findViewById(R.id.restaurantname);
        error = findViewById(R.id.error);
        industry_int = 2;

        onRegisterClick();
        onCancelClick();
    }

    // Sends required information to API and loads the next screen.
    // Error if information is invalid or incorrect
    private void sendRegistrationInfo() {
        String url = "https://capstone.api.roopairs.com/v0/auth/register/";

        HashMap<String, Object> params = new HashMap<>();
        params.put("first_name", firstname.getText().toString());
        params.put("last_name", lastname.getText().toString());
        params.put("email", email.getText().toString());
        params.put("password", password.getText().toString());

        HashMap<String, Object> internal_client = new HashMap<>();
        internal_client.put("name", restaurantname.getText().toString());
        internal_client.put("industry_type", industry_int);

        params.put("internal_client", internal_client);

        Function<JSONObject,Void> responseFunc = (jsonArray) -> {
            startActivity(new Intent(Registration.this, LocationLogin.class));
            return null;
        };

        Function<String, Void> errorFunc = (string) -> {
            error.setText("Email is already registered with a Roopairs account.");
            return null;
        };

        requestPostJsonObj(url, params, responseFunc, errorFunc, false);
    }

    // validates password
    public boolean isValidPassword(String userPassword){
        Pattern digitCasePattern = Pattern.compile("[0-9 ]");

        return !(userPassword.length() < 6) &&
                digitCasePattern.matcher(userPassword).find();
    }

    // validates email
    public boolean isValidEmail(String email){
        Pattern emailPattern =
                Pattern.compile("^[a-zA-Z0-9_+&*-]+(?:\\." +
                        "[a-zA-Z0-9_+&*-]+)*@" +
                        "(?:[a-zA-Z0-9-]+\\.)+[a-z" +
                        "A-Z]{2,7}$");

        return emailPattern.matcher(email).find();
    }

    // makes sure all fields are filled out
    private boolean isFilled() {
        return !firstname.getText().toString().isEmpty() &&
                !lastname.getText().toString().isEmpty() &&
                !email.getText().toString().isEmpty() &&
                !password.getText().toString().isEmpty() &&
                !restaurantname.getText().toString().isEmpty();
    }

    private void onRegisterClick() {
        registerButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                error.setText("");

                if (isFilled()){
                    if(isValidEmail(email.getText().toString()))
                        if(isValidPassword(password.getText().toString()))
                            sendRegistrationInfo();
                        else
                            error.setText("Invalid password.");
                    else
                        error.setText("Invalid email address.");
                }
                else
                    error.setText("Please fill out all fields.");
            }
        });
    }

    private void onCancelClick() {
        cancelButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(Registration.this, Landing.class));
            }
        });
    }
}

File: Repairs/app/src/main/java/com/rooio/repairs/Registration.kt
package com.rooio.repairs

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.*
import androidx.arch.core.util.Function
import com.android.volley.Request
import org.json.JSONException
import org.json.JSONObject
import java.util.*
import java.util.regex.Pattern

class Registration : RestApi() {

    private lateinit var firstName: EditText
    private lateinit var lastName: EditText
    private lateinit var email: EditText
    private lateinit var password: EditText
    private lateinit var passwordVerify: EditText
    private lateinit var restaurantName: EditText
    private lateinit var registerButton: Button
    private lateinit var cancelButton: Button
    private lateinit var errorMessage: TextView
    private lateinit var loadingPanel: ProgressBar
    private var industryInt: Int = 2

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_registration)
        onResume()
        centerTitleBar()
        initializeVariables()
        onRegister()
        onCancel()
        onPause()
    }

    //Initializes UI variables
    private fun initializeVariables() {
        registerButton = findViewById(R.id.register)
        cancelButton = findViewById(R.id.cancelRegistration)
        firstName = findViewById(R.id.firstName)
        firstName.requestFocus()
        lastName = findViewById(R.id.lastName)
        email = findViewById(R.id.email)
        password = findViewById(R.id.password)
        passwordVerify = findViewById(R.id.verifyPassword)
        restaurantName = findViewById(R.id.restaurantName)
        errorMessage = findViewById(R.id.errorMessage)
        loadingPanel = findViewById(R.id.loadingPanel)
    }
    // Sends required information to API and loads the next screen.
    // Error if information is invalid or incorrect
    private fun sendRegistrationInfo() {
        val url = "auth/register/"
        val params = HashMap<Any?, Any?>()
        params["first_name"] = firstName.text.toString()
        params["last_name"] = lastName.text.toString()
        params["email"] = email.text.toString()
        params["password"] = password.text.toString()
        val internalClient = HashMap<String, Any?>()
        internalClient["name"] = restaurantName.text.toString()
        internalClient["industry_type"] = industryInt
        params["internal_client"] = internalClient
        loadingPanel.visibility = View.VISIBLE
        registerButton.visibility = View.GONE

        requestJson(Request.Method.POST, JsonType.OBJECT, JsonRequest(false, url, params,
                responseFunc, errorFunc, false))
    }

    val responseFunc = Function<Any, Void?> { response: Any ->
        loadingPanel.visibility = View.GONE
        registerButton.visibility = View.VISIBLE
        val jsonObj = response as JSONObject
        storeToken(jsonObj)
        startActivity(Intent(this@Registration, LocationLogin::class.java))
        null
    }

    val errorFunc = Function<String, Void?> { error: String? ->
        loadingPanel.visibility = View.GONE
        registerButton.visibility = View.VISIBLE
        if (error == "Does not exist.") errorMessage.setText(R.string.error_register)
        else errorMessage.text = error
        null
    }

    //Stores the username and password in system
    @Throws(JSONException::class)
    private fun storeToken(responseObj: JSONObject) {
        val token = responseObj["token"] as String
        val name = responseObj["first_name"] as String

        val prefs = getSharedPreferences("UserData", Context.MODE_PRIVATE)
        val editor = prefs.edit()
        editor.putString(employeeId + "__token", token)
        editor.putString(employeeId + "__name", name)

        editor.apply()

        userToken = token
        userName = name
    }

    // validates password
    private fun isValidPassword(passwordVer: String, userPassword: String): Boolean {
        if (passwordVer == userPassword){
            val digitCasePattern = Pattern.compile("[0-9 ]")
            return userPassword.length >= 6 &&
                    digitCasePattern.matcher(userPassword).find()
        }
        else return false
    }

    // validates email
    private fun isValidEmail(email: String): Boolean {
        val emailPattern = Pattern.compile("^[a-zA-Z0-9_+&*-]+(?:\\." +
                "[a-zA-Z0-9_+&*-]+)*@" +
                "(?:[a-zA-Z0-9-]+\\.)+[a-z" +
                "A-Z]{2,7}$")
        return emailPattern.matcher(email).find()
    }

    // makes sure all fields are filled out
    private val isFilled: Boolean
        get() = firstName.text.toString().isNotEmpty() &&
                lastName.text.toString().isNotEmpty() &&
                email.text.toString().isNotEmpty() &&
                password.text.toString().isNotEmpty() &&
                passwordVerify.text.toString().isNotEmpty() &&
                restaurantName.text.toString().isNotEmpty()

    private fun onRegister() {
        loadingPanel.visibility = View.GONE
        registerButton.visibility = View.VISIBLE
        errorMessage.text = ""
        registerButton.setOnClickListener {
            if (isFilled)
                if (isValidEmail(email.text.toString()))
                    if (isValidPassword(passwordVerify.text.toString(), password.text.toString())) {
                        sendRegistrationInfo()
                    }
                    else errorMessage.text = "Invalid password."
                else errorMessage.text = "Invalid email address."
            else errorMessage.text = "Please fill out all fields."
        }
    }

    //When the user presses the cancel button to return to the landing page
    private fun onCancel() {
        cancelButton.setOnClickListener { startActivity(Intent(this@Registration, Landing::class.java)) }
    }
}